docs/contribute/testing.adoc

```diff
@@ -0,0 +1,423 @@
+= Writing and running tests for OpenSCAP
+
+This document should help you with writing and running tests for your pull
+requests. It is recommended to add a new test for every new functionality or
+if you are updating the code which is not covered by any test. Another
+recommendation is to add your test in a separate commit.
+
+All the tests reside in the link:../../tests[tests] directory. It has multiple
+subdirectories which should represent various parts of the OpenSCAP library and
+its utilities. When you contribute to some part of the OpenSCAP project you
+should put a test for your contribution into the corresponding subdirectory
+in the link:../../tests[tests] directory.
+
+It would be too time consuming to describe all the parts of the OpenSCAP
+library, its utilities and the test directories corresponding to these parts.
+Therefore, use your best judgement when deciding where to put the test for your
+pull request and if you are not sure don't be affraid to ask in the pull
+request, someone will definitely help you with that.
+
+
+== Preparing test environment and running tests
+To run a specific test or all tests you first need to compile the OpenSCAP
+library and then install additional packages required for testing. See the
+*Compilation* section in the link:../../README.md[README.md] for more details.
+
+
+== Writing a new test
+In this guide we will use an example to describe the process of writing a test
+for a pull request. Let's suppose that you already have a commit which adds
+a new feature. Let's suppose that this new feature allows to evaluate a single
+rule from a specific profile instead of all the rules in that profile.
+
+NOTE: The single rule evaluation feature was added into the OpenSCAP through
+the commit
+link:https://github.com/OpenSCAP/openscap/commit/4637b132e710d7ab73601718a531a8587b406cd3[4637b13].
+
+To sum up, we are adding the highlighted option to the `xccdf` module of
+the OpenSCAP command line tool:
+[[app-listing]]
+[subs=+quotes]
+----
+$ oscap xccdf eval *--rule X* --profile Y \
+    /usr/share/xml/scap/ssg/content/ssg-fedora-ds.xml
+----
+
+
+=== Deciding where to put a new test
+In our example, we are changing the `xccdf` module, therefore we will look for
+its subdirectory in the link:../../tests[tests] direcotory and we will find it
+at the following link: link:../../tests/API/XCCDF[tests/API/XCCDF]. There are
+multiple other subdirectories with tests and we need to decide where to put our
+test for the evaluation of a single rule. From all these subdirectories the
+most suitable one appears to be the
+link:../../tests/API/XCCDF/unittests[tests/API/XCCDF/unittests] directory as
+all the general XCCDF tests for `oscap xccdf eval` are located there. We will
+add our new test into this subdirectory.
+
```

Comment:
Do you think we could mention that our unit tests are not unit tests? If somebody decides where to put a new test it could be confusing for him if he thinks that unit tests are unit tests.

Also, it would be better to rename the tests that are not unit tests to something else than `unittests` in the repository, but that is out of scope of this PR.

Reply:
@jan-cerny I think it would be better to rename the folder rather than documenting it. I can do it in this PR, what do you say about changing it from `unittests` to `general`?

Reply:
@matusmarhefka I think this will cause problems when merging maint-1.2 to master. I prefer do not do that in this PR. We can do that in a separate PR later. Such a huge renaming might need further discussion.